File: r2r/main/api/client/ingestion.py
import json
import os
from contextlib import ExitStack
from typing import Optional, Union
from uuid import UUID

from r2r.base import ChunkingConfig


class IngestionMethods:

    @staticmethod
    async def ingest_files(
        client,
        file_paths: list[str],
        document_ids: Optional[list[UUID]] = None,
        versions: Optional[list[str]] = None,
        metadatas: Optional[list[dict]] = None,
        chunking_config_override: Optional[Union[dict, ChunkingConfig]] = None,
    ) -> dict:
        """
        Ingest files into the system.

        Args:
            file_paths (List[str]): List of file paths to ingest.
            metadatas (Optional[List[dict]]): List of metadata dictionaries for each file.
            document_ids (Optional[List[Union[UUID, str]]]): List of document IDs.
            versions (Optional[List[str]]): List of version strings for each file.
            chunking_config_override (Optional[Union[dict, ChunkingConfig]]): Custom chunking configuration.

        Returns:
            dict: Ingestion results containing processed, failed, and skipped documents.
        """
        all_file_paths = []
        for path in file_paths:
            if os.path.isdir(path):
                for root, _, files in os.walk(path):
                    all_file_paths.extend(
                        os.path.join(root, file) for file in files
                    )
            else:
                all_file_paths.append(path)

        with ExitStack() as stack:
            files = [
                (
                    "files",
                    (
                        os.path.basename(file),
                        stack.enter_context(open(file, "rb")),
                        "application/octet-stream",
                    ),
                )
                for file in all_file_paths
            ]

            data = {
                "metadatas": json.dumps(metadatas) if metadatas else None,
                "document_ids": (
                    json.dumps([str(doc_id) for doc_id in document_ids])
                    if document_ids
                    else None
                ),
                "versions": json.dumps(versions) if versions else None,
                "chunking_config_override": (
                    json.dumps(
                        chunking_config_override.model_dump()
                        if isinstance(chunking_config_override, ChunkingConfig)
                        else chunking_config_override
                    )
                    if chunking_config_override
                    else None
                ),
            }
            return await client._make_request(
                "POST", "ingest_files", data=data, files=files
            )

    @staticmethod
    async def update_files(
        client,
        file_paths: list[str],
        document_ids: Optional[list[UUID]] = None,
        metadatas: Optional[list[dict]] = None,
        chunking_config_override: Optional[Union[dict, ChunkingConfig]] = None,
    ) -> dict:
        """
        Update existing files in the system.

        Args:
            file_paths (List[str]): List of file paths to update.
            document_ids (Optional[List[str]): An optional list of document IDs to update.
            metadatas (Optional[List[dict]]): List of updated metadata dictionaries for each file.
            chunking_config_override (Optional[Union[dict, ChunkingConfig]]): Custom chunking configuration.

        Returns:
            dict: Update results containing processed, failed, and skipped documents.
        """
        if document_ids and len(file_paths) != len(document_ids):
            raise ValueError(
                "Number of file paths must match number of document IDs."
            )

        with ExitStack() as stack:
            files = [
                (
                    "files",
                    (
                        os.path.basename(file),
                        stack.enter_context(open(file, "rb")),
                        "application/octet-stream",
                    ),
                )
                for file in file_paths
            ]

            data = {
                "metadatas": json.dumps(metadatas) if metadatas else None,
                "document_ids": (
                    json.dumps([str(doc_id) for doc_id in document_ids])
                    if document_ids
                    else None
                ),
                "chunking_config_override": (
                    json.dumps(
                        chunking_config_override.model_dump()
                        if isinstance(chunking_config_override, ChunkingConfig)
                        else chunking_config_override
                    )
                    if chunking_config_override
                    else None
                ),
            }
            return await client._make_request(
                "POST", "update_files", data=data, files=files
            )

File: go/sdk/pkg/sdk/ingestion.go
package sdk

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"os"
	"path/filepath"
)

type Ingestion struct {
	client *Client
}

// IngestFiles ingests files into your R2R deployment.
//
// Parameters:
//
//	filePaths: List of file paths to ingest.
//	metadatas: List of metadata dictionaries for each file.
//	documentIDs: List of document IDs.
//	versions: List of version strings for each file.
//	chunkingConfigOverride: Custom chunking configuration.
//
// Returns:
//
//	    A map containing the response from the server.
//		 An error if the request fails, nil otherwise.
func (i *Ingestion) IngestFiles(
	filePaths []string,
	metadatas []map[string]interface{},
	documentIDs []string,
	versions []string,
	chunkingConfigOverride map[string]interface{},
) (map[string]interface{}, error) {
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)

	for _, path := range filePaths {
		file, err := os.Open(path)
		if err != nil {
			return nil, fmt.Errorf("error opening file %s: %w", path, err)
		}
		defer file.Close()

		part, err := writer.CreateFormFile("files", filepath.Base(path))
		if err != nil {
			return nil, fmt.Errorf("error creating form file: %w", err)
		}
		_, err = io.Copy(part, file)
		if err != nil {
			return nil, fmt.Errorf("error copying file content: %w", err)
		}
	}

	if metadatas != nil {
		metadatasJSON, err := json.Marshal(metadatas)
		if err != nil {
			return nil, fmt.Errorf("error marshaling metadatas: %w", err)
		}
		writer.WriteField("metadatas", string(metadatasJSON))
	}

	if documentIDs != nil {
		documentIDsJSON, err := json.Marshal(documentIDs)
		if err != nil {
			return nil, fmt.Errorf("error marshaling document IDs: %w", err)
		}
		writer.WriteField("document_ids", string(documentIDsJSON))
	}

	if versions != nil {
		versionsJSON, err := json.Marshal(versions)
		if err != nil {
			return nil, fmt.Errorf("error marshaling versions: %w", err)
		}
		writer.WriteField("versions", string(versionsJSON))
	}

	if chunkingConfigOverride != nil {
		chunkingConfigJSON, err := json.Marshal(chunkingConfigOverride)
		if err != nil {
			return nil, fmt.Errorf("error marshaling chunking config: %w", err)
		}
		writer.WriteField("chunking_settings", string(chunkingConfigJSON))
	}

	err := writer.Close()
	if err != nil {
		return nil, fmt.Errorf("error closing multipart writer: %w", err)
	}

	result, err := i.client.makeRequest("POST", "ingest_files", body, writer.FormDataContentType())
	if err != nil {
		return nil, err
	}

	stats, ok := result.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: expected map[string]interface{}, got %T", result)
	}

	return stats, nil
}

// UpdateFiles updates existing files in your R2R deployment.
//
// Parameters:
//
//	filePaths: List of file paths to update.
//	documentIDs: List of document IDs.
//	metadatas: List of metadata dictionaries for each file.
//	chunkingConfigOverride: Custom chunking configuration.
//
// Returns:
//
//	    A map containing the response from the server.
//		 An error if the request fails, nil otherwise.
func (i *Ingestion) UpdateFiles(
	filePaths []string,
	documentIDs []string,
	metadatas []map[string]interface{},
	chunkingConfigOverride map[string]interface{},
) (map[string]interface{}, error) {
	if len(filePaths) != len(documentIDs) {
		return nil, fmt.Errorf("number of file paths must match number of document IDs")
	}

	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)

	for _, path := range filePaths {
		file, err := os.Open(path)
		if err != nil {
			return nil, fmt.Errorf("error opening file %s: %w", path, err)
		}
		defer file.Close()

		part, err := writer.CreateFormFile("files", filepath.Base(path))
		if err != nil {
			return nil, fmt.Errorf("error creating form file: %w", err)
		}
		_, err = io.Copy(part, file)
		if err != nil {
			return nil, fmt.Errorf("error copying file content: %w", err)
		}
	}

	documentIDsJSON, err := json.Marshal(documentIDs)
	if err != nil {
		return nil, fmt.Errorf("error marshaling document IDs: %w", err)
	}
	writer.WriteField("document_ids", string(documentIDsJSON))

	if metadatas != nil {
		metadatasJSON, err := json.Marshal(metadatas)
		if err != nil {
			return nil, fmt.Errorf("error marshaling metadatas: %w", err)
		}
		writer.WriteField("metadatas", string(metadatasJSON))
	}

	if chunkingConfigOverride != nil {
		chunkingConfigJSON, err := json.Marshal(chunkingConfigOverride)
		if err != nil {
			return nil, fmt.Errorf("error marshaling chunking config: %w", err)
		}
		writer.WriteField("chunking_settings", string(chunkingConfigJSON))
	}

	err = writer.Close()
	if err != nil {
		return nil, fmt.Errorf("error closing multipart writer: %w", err)
	}

	result, err := i.client.makeRequest("POST", "update_files", body, writer.FormDataContentType())
	if err != nil {
		return nil, err
	}

	stats, ok := result.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: expected map[string]interface{}, got %T", result)
	}

	return stats, nil
}
